File: assets/src/shared/lib/crypto/hash.ts
import { blake3 } from "@noble/hashes/blake3.js";
import { decodeBase64UrlStrict, encodeBase64Url } from "./encoding";

export type HashSentinelPolicy = "none" | ReadonlySet<string>;

const BLAKE3_B64URL_RE = /^[A-Za-z0-9_-]{43}$/;

export function blake3Base64Url(bytes: Uint8Array): string {
  return encodeBase64Url(blake3(bytes));
}

export function assertBlake3Base64Url(
  value: string,
  sentinelPolicy: HashSentinelPolicy = "none",
): void {
  if (sentinelPolicy !== "none" && sentinelPolicy.has(value)) {
    return;
  }
  if (!BLAKE3_B64URL_RE.test(value)) {
    throw new Error("invalid_blake3_base64url");
  }
  decodeBase64UrlStrict(value, 32);
  if (/^[0-9a-f]{64}$/i.test(value)) {
    throw new Error("invalid_blake3_hex");
  }
}

File: lib/refmd/crypto/hash.ex
defmodule RefMD.Crypto.Hash do
  @moduledoc false

  alias RefMD.Crypto.Blake3
  alias RefMD.Crypto.Encoding

  @blake3_b64url_re ~r/^[A-Za-z0-9_-]{43}$/
  @hex_re ~r/^[0-9a-fA-F]{64}$/

  def blake3_base64url(bytes) when is_binary(bytes), do: Blake3.hash_base64url(bytes)

  def assert_blake3_base64url!(value), do: assert_blake3_base64url!(value, :none)

  def assert_blake3_base64url!(value, sentinel_policy) when is_binary(value) do
    if sentinel_policy != :none && MapSet.member?(sentinel_policy, value) do
      :ok
    else
      cond do
        Regex.match?(@hex_re, value) ->
          raise ArgumentError, "invalid_blake3_hex"

        not Regex.match?(@blake3_b64url_re, value) ->
          raise ArgumentError, "invalid_blake3_base64url"

        true ->
          Encoding.decode_base64url!(value, 32)
          :ok
      end
    end
  end
end
